e2e/test/helpers/RetryOperationHelper.cs

```diff
@@ -47,19 +56,28 @@ public static async Task RetryOperationsAsync(Func<Task> asyncOperation, IRetryP
                 }
 
                 logger.Trace($"Will retry operation in {retryInterval}.");
-                await Task.Delay(retryInterval).ConfigureAwait(false);
+                await Task.Delay(retryInterval, cancellationToken);
             }
-            while (shouldRetry);
+            while (shouldRetry && !cancellationToken.IsCancellationRequested);
```

Comment:
Probably shouldn't just end the loop without some kind of exception that says why the operation was cancelled.

I think for the cancellation token being signalled, that is handled in line 59 - it'll produce an OperationCancelledException.

But what about `shouldRetry` being false?

Reply:
When `shouldRetry` is false, the retriable `IotHubServiceException` will be thrown out and we log that as well.